### ISSUE TITLE
Tracking issue for RFC 2093: Infer `T: 'x` outlives requirements on structs 

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#44493]
This is a tracking issue for the RFC "Infer `T: 'x` outlives requirements on structs " (rust-lang/rfcs#2093).

**Current status:**

Implemented and we have **decided to stabilize**. We still need someone to make the stabilization PR! Mentoring instructions here:

https://github.com/rust-lang/rust/issue_comments#issuecomment-411781121

Until then, you can use this by adding `#![feature(infer_outlives_requirements)]` to your crate.

**Steps:**

- [x] Implement the RFC ([mentoring instructions here](https://github.com/rust-lang/rust/issues/44493#issuecomment-330279870))
- [ ] Adjust documentation ([see instructions on forge][doc-guide])
- [ ] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

**Unresolved questions:**

The interaction with `'static` remains a bit unclear. For example, do we want to infer a `T: 'static` requirement on the type parameter `T` here? The code [currently will do so](https://play.rust-lang.org/?gist=1b6ba427cbced22f3881a8fda733ab24&version=nightly&mode=debug), arguably leading to confusing errors.

```rust
#![feature(infer_outlives_requirements)]

struct Foo<T> { t: &'static T }

fn main() {
    let x = 3;
    let _ = Foo { t: &x };
}
```

**Comment by nikomatsakis** [comment]
### Mentoring instructions

In the compiler, when we want to know what predicates are defined on something (e.g., a struct), we do that via the `predicates_of` query. Right now, this is implemented by [this code in `librusc_typeck`](https://github.com/rust-lang/rust/blob/3a7b960731fd1cf86f1879b1a0d44196a0917c53/src/librustc_typeck/collect.rs#L1354-L1356). This code basically reads the HIR and produces *exactly* the results found there.

In terms of how to fit this inference into the compiler pipeline, then, I think we want to ensure that `predicates_of` includes these inferred predicates. To do that, we can redefine `predicates_of` in terms of two new queries, `explicit_predicates_of` and `inferred_outlives_of` -- basically, `predicates_of` would be the union of the two:

```
predicates_of(D) = explicit_predicates_of(D) + inferred_outlives_of(D)
```

`explicit_predicates_of` would be exactly the same as `predicates_of` is today (and only defined for local crates).  It just returns the predicates the user typed. `inferred_outlives_of`, in contrast, will include the predicates that we are going to infer.

**The very first PR, then, could be to introduce the `explicit_predicates_of` query and to make `predicates_of` just redirect to it** (later, we will add `inferred_outlives_of`). This makes sense because we're going to need `explicit_predicates_of` when defining `inferred_outlives_of`.

Now we have to figure out how to implemented `inferred_outlives_of`. I think the basic structure here is going to be similar to how [variance inference][] is setup. In terms of queries, variance inference is defined by [two queries][]:

[variance inference]: https://github.com/rust-lang/rust/tree/3a7b960731fd1cf86f1879b1a0d44196a0917c53/src/librustc_typeck/variance
[two queries]: https://github.com/rust-lang/rust/blob/3a7b960731fd1cf86f1879b1a0d44196a0917c53/src/librustc/ty/maps.rs#L1214-L1220

- `crate_variances` -- returns a map containing the variance for every item in the crate
- `variances_of` -- just selects an item of this map.

The `crate_variances` query is intended as an implementation detail of variance -- end-users should use `variances_of` for a particular item.

The reason for this particular setup is that we can't infer the variances for a **single item** in isolation; the variance for an item X depends on the contents of the struct, and there may be cycles. Since cycles are generally forbidden between queries, we instead compute the variances for ALL structs in the crate, and then use individual queries to extract the result.

So, for `inferred_outlives_of`, we will basically create a new query, probably living alongside `variance` in `librustc_typeck`, let's call it `inferred_outlives_of`. This query will defer to a crate-wide query `inferred_outlives_crate` (we should probably have named `crate_variances` as `variances_crate`...).  Since we only infer predicates for struts, `inferred_outlives_of` can just return the empty set for any def-id that is not a struct.

This crate-wide computation will work much as it is [described in the RFC][algo] -- there will be a set of constraints being inferred for each struct `S`, what the RFC calls `A[S]`. These sets will be  sets of [`Predicate<'tcx>`] values. We will initialize them by filtering the `explicit_predicates_of` each struct to only include those of type [`TypeOutlives`] or [`RegionOutlives`].

[`Predicate<'tcx>`]: https://github.com/rust-lang/rust/blob/3a7b960731fd1cf86f1879b1a0d44196a0917c53/src/librustc/ty/mod.rs#L816
[`TypeOutlives`]: https://github.com/rust-lang/rust/blob/3a7b960731fd1cf86f1879b1a0d44196a0917c53/src/librustc/ty/mod.rs#L828-L829
[`RegionOutlives`]: https://github.com/rust-lang/rust/blob/3a7b960731fd1cf86f1879b1a0d44196a0917c53/src/librustc/ty/mod.rs#L825-L826
[algo]: https://github.com/rust-lang/rfcs/blob/master/text/2093-infer-outlives.md#reference-level-explanation

Next, we can walk the types of the fields declared in each struct. These can be obtained by invoking `type_of(def_id)` where `def_id` is the `DefId` of the struct. This should give back a type of the `TyAdt` variant; from this, you can extract the types of the fields. [Something like this code from variance](https://github.com/rust-lang/rust/blob/3a7b960731fd1cf86f1879b1a0d44196a0917c53/src/librustc_typeck/variance/constraints.rs#L182-L194) should give you the idea. For each field that references another struct `S2`, we need to create a link between those types and ensure that their implied predicates are a superset.

Once this is done, we should have some kind of inferred set of obligations for each struct. I think it'd be good to setup a unit-testing mechanism for this similar to [the one we use for variance](https://github.com/rust-lang/rust/blob/3a7b960731fd1cf86f1879b1a0d44196a0917c53/src/librustc_typeck/variance/test.rs). Basically, some custom code that looks for a `#[rustc_inferred_outlives]` annotation and, when found, dumps out a "compilation error" that includes the results. This lets us write unit tests like [this one](https://github.com/rust-lang/rust/blob/3a7b960731fd1cf86f1879b1a0d44196a0917c53/src/test/compile-fail/variance-associated-types.rs), where we just check the result of this query.

So, in terms of first steps, these are good PRs to open:

- Check `predicates_of` to be implemented in terms of a subsidiary predicate `explicit_predicates_of`.
- Introduce `inferred_outlives` predicate and unit-testing mechanism. Initially, the predicate can always return an empty set of predicates, but we can add some tests nonetheless; they'll just initially show an empty result for the predicate.
- Implement the predicate inference, fix the test, add more tests.

**Comment by toidiu** [comment]
I would like to work on this!

**Comment by nikomatsakis** [comment]
@toidiu hey, just checking in! I just r+'d that first PR, wondering if you'd had any time to mess around with the next few steps?

**Comment by toidiu** [comment]
@nikomatsakis hey sry for the slow progress. I have have mainly been trying to wrap my head around all the different concepts and the repo structure. Have read the mentoring guide and rfc a few times now I am hoping to make some initial pr's early this week.

**Comment by nikomatsakis** [comment]
Pseudocode of how the inference wants to work, I think:

```
// Compute a map from each struct/enum/union S to the **explicit**
// outlives predicates (`T: 'a`, `'a: 'b`) that the user wrote.
// Typically there won't be many of these, except in older code where
// they were mandatory. Nonetheless, we have to ensure that every such
// predicate is satisfied, so they form a kind of base set of requirements
// for the type.
let mut explicit_outlives_predicates = map();
for def_id in all_types() {
    let explicit_predicates = tcx.explicit_predicates(def_id);
    let filtered_predicates = explicit_predicates.iter().filter_map(is_outlives_predicate).collect();
    explicit_outlives_predicates.insert(
        def_id,
        filtered_predicates);
}

// Create the sets of inferred predicates for each type. These sets
// are initially empty but will grow during the inference step.
let mut inferred_outlives_predicates = map();
for def_id in all_types() {
    inferred_outlives_predicates.insert(def_id, empty_set());
}

// Now comes the inference part. We iterate over each type and figure out,
// for each of its fields, what outlives predicates are needed to make that field's
// type well-formed. Those get added to the `inferred_outlives_predicates` set
// for that type.
let mut changed = true;
while changed {
    changed = false;
    for def_id in all_types() {
        for field_ty in the HIR type definition {
            let required_predicates = required_predicates_for_type_to_be_wf(field_ty);
            inferred_outlives_predicates.extend(required_predicates);
            if new predicates were added { changed = true; }
        }
    }
}

// When we are done, `inferred_outlives_predicates` is the resulting map.
```

**Comment by Yoric** [comment]
@toidiu Are you still working on this? If not, I'd be interested in taking a jab at this issue.

**Comment by toidiu** [comment]
@Yoric yea still making progress and iterating on one PR so as not to queue up bors

**Comment by nikomatsakis** [comment]
An interesting question arose during the course of the implementation:

The current implementation will happily infer a `T: 'static` bound if one arises. So for example:

```rust
trait Foo<T: 'static> { .. }
struct Bar<T> { b: Box<dyn Foo<T>> }
```

`Bar<T>` here will have an inferred `T: 'static` bound. This seems to me to be quite a bit more surprising than bounds that arise between lifetimes declared on the struct itself (e.g., the classic case of `struct Foo<'a, T>(&'a T)`).

I think I'm inclined not to infer `T: 'static` to start and instead require it to be manually added.

That said, I think a similar argument can be made that this will not cause a big problem in practice: in particular, in order to construct `Bar`, one must construct a trait object, and of course one cannot make the trait object unless `T: 'static`.

**Comment by mark-i-m** [comment]
@nikomatsakis What if `T: 'static` is inferred for `Bar<T>` too, but it is _also_ included in the rustdocs? I think as long as it is documented, it shouldn't be a problem, right?

**Comment by mark-i-m** [comment]
Actually, when I thought again, I'm not so sure it's a good idea: what impact does this have on private dependencies? In other words if Foo is in your crate and Bar is in mine, if you add a bound to Foo, it becomes a breaking change in _both_ crates without me having to change my version number!

Perhaps one strategy is to not infer bounds across crate boundaries?

**Comment by nikomatsakis** [comment]
@mark-i-m adding a bound to `Foo` ought to be a breaking change for the `foo` crate, so `bar` would not get it "by default".

**Comment by nikomatsakis** [comment]
Though it's certainly subtle: one might fail to recognize that you had made a semver breaking change.

**Comment by joshtriplett** [comment]
@aturon 
> The interaction with 'static remains a bit unclear. For example, do we want to infer a T: 'static requirement on the type parameter T here? The code currently will do so, arguably leading to confusing errors.

Is this still true? If so, should we be stabilizing? If not, could you please update the summary comment at the top to clarify the current state?

**Comment by toidiu** [comment]
This is the relevant line of code.
https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/outlives/utils.rs#L143

The current implementation will infer `T: 'static`

**Comment by joshtriplett** [comment]
@rfcbot concern inferred-static

We need to decide if the current behavior of inferring `'static` is correct, and either document it or change it.

**Comment by aturon** [comment]
Nominating for Lang Team discussion.

**Comment by nikomatsakis** [comment]
So regarding @joshtriplett's concern —

I personally think we could move that to a separate feature gate, just to be cautious. It does feel mildly different to infer `T: 'a` between the parameters on a struct vs inferring the requirement that `T: 'static`, though I'm having trouble putting my finger on why.

That said, many of the same arguments do seem to apply --- e.g., if you can construct the struct, then you must have a static reference to the data in question, and so forth.

As to displaying these bounds in rustdoc, I don't know. I wonder if it would be nice to have a "fully elaborated" view in rustdoc than one can enable, which would also show the inferred variance and so forth (that would require us to decide whether to say that `'a` is co- or contra-variant in `&'a T`, something which till now has been a kind of 'implementation detail'). I feel like I would prefer not to confront the user with the full detail by default, but I can imagine it being useful to be able to see it.

I'd like to try deploying this feature throughout rustc though, I've not had a chance to do that yet.

**Comment by joshtriplett** [comment]
@nikomatsakis A separate feature gate would certainly unblock this.

**Comment by toidiu** [comment]
I'll put up a PR to hide `T: 'static` behind a feature gate.

**Comment by scottmcm** [comment]
#51858 appears to be closed, so I'll assume the OP's "blocking problems" list is just out of date.

**Comment by joshtriplett** [comment]
@rfcbot resolved inferred-static

Addressed by https://github.com/rust-lang/rust/pull/52761

**Comment by nikomatsakis** [comment]
Let's get this stabilized, then!

There are instructions for how to do so here:

https://forge.rust-lang.org/stabilization-guide.html

@toidiu — you would be up for that? You've seen this thing through from start to finish thus far!

**Comment by nikomatsakis** [comment]
I think we should also split off a new issue for the `'static` feature gate...

**Comment by toidiu** [comment]
@nikomatsakis yep would be glad to :)

**Comment by pnkfelix** [comment]
visiting for triage. PR #53793 is meant to resolve this.

**Comment by pnkfelix** [comment]
visiting for triage. #53793 landed. closing as fixed.

**Comment by nikomatsakis** [comment]
Created https://github.com/rust-lang/rust/issues/54185 to track the `'static` question

**Comment by nikomatsakis** [comment]
https://github.com/rust-lang/rust/issues/52042 already exists to track an idiom lint

**Comment by pnkfelix** [comment]
https://github.com/rust-lang/rust/issues/54185#issuecomment-1124097663 ended the question about whether to special-case `'static` and removed `#![feature(infer_static_outlives_requirements)]`.
